server/aesdsocket.c

```diff
@@ -77,52 +108,236 @@ void cleanUp(void)
+    pthread_mutex_destroy(&trasactionMutex);
+    
+
+    free(thisNode);
+    
+
+    syslog(LOG_INFO,"Caught signal or failed, exiting");
 
     closelog();
     remove(DATA_PATH);
+    exit(exitVal);
 }
 
 void exitSigHandler(int signo) 
 {
     if (signo == SIGINT || signo == SIGTERM) 
     {
-        cleanUp();
-        exit(0);
+        // Setup for main to cleanup!
+        cleanUpTime = true;
+        cleanUp(EXIT_SUCCESS); // not totally safe but passes for now... The problem is cleaningup time stamper quickly for the test
+    }
+}
+
+void *timeStamper () 
+{
+    // chat gpt "This program appends a timestamp in the form “timestamp:time” to the /var/tmp/aesdsocketdata 
+    // file every 10 seconds using the RFC 2822 compliant strftime format. The string includes the year, month, 
+    // day, hour (in 24 hour format), minute, and second representing the system wall clock time."
+    while(!cleanUpTime)
+    {
+        char timestamp[128] ;
+        sleep(DELAY_TO_STAMP);
+
+        pthread_mutex_lock(&trasactionMutex);
+        time_t t = time(NULL);
+        struct tm *tm = localtime(&t);
+        int timeStampfp;
+        strftime(timestamp, sizeof(timestamp), "timestamp: %Y-%m-%d %H:%M:%S\n", tm);
+        timeStampfp = open(DATA_PATH,   O_RDWR | O_CREAT |  O_APPEND, 0664);
+        if(write(timeStampfp,timestamp,sizeof(timestamp)) == -1)
+        {
+            syslog(LOG_INFO,"Failed to write timeStamp");
+            cleanUp(EXIT_FAILURE);
+        }     
+        close(timeStampfp);
+        tm = NULL;
+        pthread_mutex_unlock(&trasactionMutex);
+
+    }
+    cleanUp(EXIT_SUCCESS);
+    return NULL;
+}
+
+
+
+
+
+
+//https://blog.taborkelly.net/programming/c/2016/01/09/sys-queue-example.html
+void *add_to_queue_and_send(void *arg) {
+    struct node* addedNode = (struct node *)arg;
+    pthread_mutex_lock(&mutex);
+    
+    
+    SLIST_INSERT_HEAD(&head, addedNode, nodes);
+    
+    
+    pthread_mutex_unlock(&mutex);
+
+
+    // Implentation is a combination of both cases of either 4 or 6!
+    //https://stackoverflow.com/questions/3060950/how-to-get-ip-address-from-sock-structure-in-c
+    char ipAccepted[INET6_ADDRSTRLEN]; // handle both Inet6 and 4
+    void *addrPtr;
+
+    // if INET 4 
+    if(client_addr.ss_family == AF_INET)
+    {
+        // then save the address
+        struct sockaddr_in *tempSock = (struct sockaddr_in *)&client_addr;
+        addrPtr = &(tempSock->sin_addr);
+    }
+    else
+    {
+        // save the INET6 address
+        struct sockaddr_in6 *tempSock = (struct sockaddr_in6 *)&client_addr;
+        addrPtr = &(tempSock->sin6_addr);
+    }
+
+    // translate the pointer to a ip using the INET_NTOP function
+    inet_ntop(client_addr.ss_family, addrPtr, ipAccepted, sizeof(ipAccepted));
+
+    syslog(LOG_INFO,"Accepted connection from %s",ipAccepted);
+
+    
+
+    
+    int bytes_received ;
+
+    bool recieving = true;
+    pthread_mutex_lock(&trasactionMutex);
```

Comment:
@Nicholas-Buckley77 this looks like it is locking a code block rather than a specific access to the write file.  As written there's not much benefit of multithreading since each thread will need to run to completion due to lock usage.